code-analysis/crates/codegen/parser/src/chumsky/boilerplate.rs

```diff
@@ -88,55 +82,69 @@ pub fn cst_head() -> TokenStream {
         }
 
         impl Node {
-            pub fn none() -> Rc<Self> {
-                Rc::new(Self::None)
+            pub fn rule(kind: kinds::Rule, children: Vec<Option<Rc<Self>>>) -> Option<Rc<Self>> {
+                let children: Vec<_> = children.into_iter().filter_map(|e| e).collect();
+                if children.is_empty() { None } else { Some(Rc::new(Self::Rule { kind, children })) }
             }
 
-            pub fn rule(kind: kinds::Rule, children: Vec<Rc<Self>>) -> Rc<Self> {
-                Rc::new(Self::Rule { kind, children })
-            }
-
-            pub fn trivia_token(kind: kinds::Token, lex_node: Rc<lex::Node>) -> Rc<Self> {
-                Rc::new(Self::Token {
+            pub fn trivia_token(kind: kinds::Token, lex_node: Rc<lex::Node>) -> Option<Rc<Self>> {
+                Some(Rc::new(Self::Token {
                     kind,
                     lex_node,
                     trivia: vec![],
-                })
+                }))
             }
 
             pub fn token(
                 kind: kinds::Token,
                 lex_node: Rc<lex::Node>,
-                leading_trivia: Rc<Self>,
-                trailing_trivia: Rc<Self>,
-            ) -> Rc<Self> {
+                leading_trivia: Option<Rc<Self>>,
+                trailing_trivia: Option<Rc<Self>>,
+            ) -> Option<Rc<Self>> {
                 let mut trivia = vec![];
-                if *leading_trivia != Self::None {
+                if let Some(leading_trivia) = leading_trivia {
                     trivia.push(leading_trivia)
                 }
-                if *trailing_trivia != Self::None {
+                if let Some(trailing_trivia) = trailing_trivia {
                     trivia.push(trailing_trivia)
                 }
-                Rc::new(Self::Token {
+                Some(Rc::new(Self::Token {
                     kind,
                     lex_node,
                     trivia,
-                })
+                }))
             }
 
-            pub fn group(children: Vec<Rc<Self>>) -> Rc<Self> {
-                if children.is_empty() {
-                    Self::none()
+            pub fn group(children: Vec<Option<Rc<Self>>>) -> Option<Rc<Self>> {
+                let children: Vec<_> = children.into_iter().filter_map(|e| e).collect();
+                if children.is_empty() { None } else { Some(Rc::new(Self::Group { children })) }
+            }
+
+            pub fn top_level_token(lex_node: Option<Rc<lex::Node>>) -> Rc<Self> {
+                if let Some(lex_node) = lex_node {
+                    if let lex::Node::Named(kind, lex_node) = lex_node.as_ref() {
+                        Rc::new(Self::Token {
+                            kind: *kind,
+                            lex_node: lex_node.clone(),
+                            trivia: vec![],
+                        })
+                    } else {
+                        unreachable!("Top level token unexpected result: {:?}", lex_node)
```

Comment:
Reading [the docs](https://doc.rust-lang.org/std/macro.unreachable.html) on `unreachable!`, it seems it is only a shorthand for `panic!`. Should we be using `expect()` instead for such cases?

Reply:
The name of the macro conveys meaning, and might in the future have a more sophisticated implementation than just panicing. IMO unreachable is the right thing to use for code that violates conditions that exist purely because of language weakness.